### PR TITLE
Add engine for monitoring AWS S3 buckets

### PR DESCRIPTION
**Pull request by tschoonj**

### CODE REVIEW

**Comment by tschoonj** [comment]
My work here is done... @LAShemilt  let me know if you have any comments, otherwise I will merge it in.

File: rfi_file_monitor/operations/dummy_operation.py [hunk 1]
```diff
@@ -9,12 +9,13 @@
...
 from ..utils.exceptions import SkippedOperation
-from ..utils.decorators import with_pango_docs
-from ..file import File
+from ..utils.decorators import with_pango_docs, supported_filetypes
+from ..file import File, RegularFile, AWSS3Object
 
 logger = logging.getLogger(__name__)
 
 @with_pango_docs(filename='dummy_operation.pango')
+@supported_filetypes(filetypes=[RegularFile, AWSS3Object])
```
Comment (LAShemilt) [inline rfi_file_monitor/operations/dummy_operation.py:18]: I guess this decorator will also need to go on the other classes, like sftp uploader... Also in the extensions

Comment (tschoonj) [reply rfi_file_monitor/operations/dummy_operation.py:18]: If they only support `RegularFile`, then the decorator is not necessary.

File: rfi_file_monitor/utils/__init__.py [hunk 2]
```diff
@@ -103,14 +104,28 @@ def get_md5(fname: os.PathLike) -> str:
             hash_md5.update(chunk)
     return hash_md5.hexdigest()
 
+def get_file_creation_timestamp(file_path: os.PathLike):
+    # get creation time, or something similar...
+    # https://stackoverflow.com/a/39501288
+    if platform.system() == 'Windows':
```
Comment (LAShemilt) [inline rfi_file_monitor/utils/__init__.py:110]: this is interesting. When did it change that getting a timestamp platform dependent?

File: rfi_file_monitor/utils/decorators.py [hunk 3]
```diff
@@ -74,19 +75,21 @@ def supported_filetypes(filetypes: Union[Type[File], Sequence[Type[File]]]):
...
+        if isinstance(filetypes, collections.abc.Sequence):
+            _filetypes = filetypes
+        else:
+            _filetypes = [filetypes]
+        for filetype in _filetypes:
             if filetype in _app.filetypes_supported_operations_map:
                 _app.filetypes_supported_operations_map[filetype].append(cls)
             else:
-                _app.filetypes_supported_operations_map[filetype] = list(cls)
+                _app.filetypes_supported_operations_map[filetype] = [cls]
```
Comment (LAShemilt) [inline rfi_file_monitor/utils/decorators.py:92]: yes this was a weird one, I also found that list() does not work but [ ] does...

File: rfi_file_monitor/utils/__init__.py (hunk 2 above)
Comment (tschoonj) [reply rfi_file_monitor/utils/__init__.py:110]: It's a bit of a mess indeed. This seems to be most reliable solution

File: rfi_file_monitor/utils/decorators.py (hunk 3 above)
Comment (tschoonj) [reply rfi_file_monitor/utils/decorators.py:92]: yep, basically `list()`, `tuple()`, `int()` etc. act as casts when they have an argument, which is not always what you want

**Comment by LAShemilt** [comment]
Hi I just ran this  and some creds might not be correct causing the error - however it causes the file monitor to hang in the preflight check and does not give me any feedback on the error. 

`
Exception in thread Thread-2:
Traceback (most recent call last):
  File "c:\users\vpf26432\pycharmprojects\rfi-file-monitor\rfi_file_monitor\engines\aws_s3_bucket_engine.py", line 279, in run
    self._engine.s3_client.head_bucket(Bucket=self._engine.params.bucket_name)
  File "C:\Users\vpf26432\Anaconda3\envs\rfi-file-monitor\lib\site-packages\botocore\client.py", line 357, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "C:\Users\vpf26432\Anaconda3\envs\rfi-file-monitor\lib\site-packages\botocore\client.py", line 676, in _make_api_call
    raise error_class(parsed_response, operation_name)
botocore.exceptions.ClientError: An error occurred (403) when calling the HeadBucket operation: Forbidden

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\vpf26432\Anaconda3\envs\rfi-file-monitor\lib\threading.py", line 950, in _bootstrap_inner
    self.run()
  File "c:\users\vpf26432\pycharmprojects\rfi-file-monitor\rfi_file_monitor\engines\aws_s3_bucket_engine.py", line 282, in run
    GLib.idle_(self._engine._abort, self._task_window, e, priority=GLib.PRIORITY_HIGH)
  File "C:\Users\vpf26432\Anaconda3\envs\rfi-file-monitor\lib\site-packages\gi\overrides\__init__.py", line 32, in __getattr__
    return getattr(self._introspection_module, name)
  File "C:\Users\vpf26432\Anaconda3\envs\rfi-file-monitor\lib\site-packages\gi\module.py", line 123, in __getattr__
    raise AttributeError("%r object has no attribute %r" % (
AttributeError: 'gi.repository.GLib' object has no attribute 'idle_'
Traceback (most recent call last):
  File "C:\Users\vpf26432\Anaconda3\envs\rfi-file-monitor\Scripts\rfi-file-monitor-script.py", line 33, in <module>
    sys.exit(load_entry_point('rfi-file-monitor', 'console_scripts', 'rfi-file-monitor')())
  File "c:\users\vpf26432\pycharmprojects\rfi-file-monitor\rfi_file_monitor\__init__.py", line 46, in main
    rv = app.run(sys.argv)
  File "C:\Users\vpf26432\Anaconda3\envs\rfi-file-monitor\lib\site-packages\gi\overrides\Gio.py", line 43, in run
    return Gio.Application.run(self, *args, **kwargs)
  File "C:\Users\vpf26432\Anaconda3\envs\rfi-file-monitor\lib\contextlib.py", line 124, in __exit__
    next(self.gen)
  File "C:\Users\vpf26432\Anaconda3\envs\rfi-file-monitor\lib\site-packages\gi\_ossighelper.py", line 237, in register_sigint_fallback
    signal.default_int_handler(signal.SIGINT, None)
`

**Comment by tschoonj** [comment]
That's a bug indeed. Thanks, I will push a fix.

**Pull request merged**